### PR TITLE
Add log inline log tags

### PR DESCRIPTION
**Pull request by amitgoldarb**

### CODE REVIEW

**Comment by amitizle** [comment]
I'm feeling like I'm missing something but I'm not sure what, I'd appreciate your insight.
There's the ["regular" pino API](https://getpino.io/#/docs/api?id=mergingobject).
In short, in `pino` your first argument can be either a string or an object. If it's an object (from the link above)
> Each enumerable key and value of the `mergingObject` is copied in to the JSON log line.

Otherwise, if the first argument is a string;
> A `message` string can optionally be supplied as the first parameter, or as the second parameter after supplying a `mergingObject`. By default, the contents of the `message` parameter will be merged into the JSON log line under the msg key.

Everything that comes **after** the message (either 1st or 2nd argument for each log line) is considered [interpolationValues](https://getpino.io/#/docs/api?id=interpolationvalues-any).

Why are we not just transporting this the same way to `pino`, and implementing the same logic?

File: packages/logging/src/internal/pino.ts [hunk 1]
```diff
@@ -71,14 +70,31 @@ const formatError = (input: FormatterInput): string => [
...
+    : ['', '']
+  )
 
+const formatExcessArgs = (excessArgs?: unknown[]): LogTags => {
+  const baseExcessArgs = (excessArgs || []).filter(x => x)
+  const formattedPrimitiveExcessArgs = Object.fromEntries(
+    // filter for primitive type inside formatPrimitiveExcessArg to maintain global index
+    baseExcessArgs
+      .map(formatPrimitiveExcessArg)
+      .filter(x => x[0])
```
Comment (amitizle) [inline packages/logging/src/internal/pino.ts:85]: Can this ever be undefined or false?

File: packages/logging/test/internal/pino_logger.test.ts [hunk 2]
```diff
@@ -669,22 +669,24 @@ describe('pino based logger', () => {
           level: 'warn',
           message: 'where is extra args object foo',
           arg0: 'moo',
-          arg1: '{"extra":"should be in log"}',
-          arg2: 'true',
+          extra: 'should be in log',
+          arg2: true,
           arg3: '"string with \\"bad chars\\"\\t\\n"',
         })
       })
```
Comment (amitizle) [inline packages/logging/test/internal/pino_logger.test.ts:676]: I think we should be adding more JSON format tests, I mean get the line, `JSON.parse` is and check that all keys are there

File: packages/logging/src/internal/log-tags.ts [hunk 3]
```diff
@@ -31,23 +32,40 @@ const stringifyIfNecessary = (s: string): string => {
   }
 }
 
-export const formatLogTagValue = (value: unknown): string => {
-  if (typeof value === 'number' || typeof value === 'boolean') return `${value.toString()}`
-  if (typeof value === 'string') return `${stringifyIfNecessary(value)}`
-  return ''
-}
+const encapsulateIfNecessary = (s: string): string => (s.match(/^".*"$/) ? s : `"${s}"`)
```
[the remark this answers is not in the document]
Comment (amitgoldarb) [reply packages/logging/src/internal/log-tags.ts:35]: I think it is better to be consistent and wrap all strings with ". Fixed the flow and stringified all strings. No more need for this function.

File: packages/logging/src/internal/log-tags.ts [hunk 4]
```diff
@@ -31,23 +32,40 @@ const stringifyIfNecessary = (s: string): string => {
...
+
+
+const formatKeyValue = (key: string, value: string): string =>
+  `${key}=${value}`
+
+const formatObjectLogTag = (value: Record<string, unknown>): string => {
+  const keys = Object.keys(value)
+  return keys
+    .map(key => formatKeyValue(key, safeStringify(value[key])))
+    .join(' ')
```
[the remark this answers is not in the document]
Comment (amitgoldarb) [reply packages/logging/src/internal/log-tags.ts:54]: Cool thanks! fixed

File: packages/logging/src/internal/log-tags.ts [hunk 5]
```diff
@@ -31,23 +32,40 @@ const stringifyIfNecessary = (s: string): string => {
...
+
 export const formatLogTags = (logTags: Record<string, unknown>, baseKeys: string[]): string => {
   const tagsWithoutBaseKeys = _.omit(logTags, ...baseKeys)
   return Object.keys(tagsWithoutBaseKeys)
-    .map(logTagKey => (
-      isPrimitiveType(logTags[logTagKey])
-        ? `${logTagKey}=${formatLogTagValue(logTags[logTagKey])}`
-        : ''
-    ))
+    .map(logTagKey => {
```
[the remark this answers is not in the document]
Comment (amitgoldarb) [reply packages/logging/src/internal/log-tags.ts:60]: Cool, fixed!

File: packages/logging/src/internal/log-tags.ts [hunk 6]
```diff
@@ -31,23 +32,40 @@ const stringifyIfNecessary = (s: string): string => {
...
+      if (isPrimitiveType(logTags[logTagKey])) {
+        return formatKeyValue(logTagKey, formatPrimitiveLogTagValue(logTags[logTagKey]))
+      }
+      if (typeof logTags[logTagKey] === 'object') {
+        return formatObjectLogTag({ [logTagKey]: logTags[logTagKey] })
+      }
+      return ''
```
[the remark this answers is not in the document]
Comment (amitgoldarb) [reply packages/logging/src/internal/log-tags.ts:67]: I think it's better to support other types implicitly. I don't want to support functions/undefined/symbol types. I think the only other one we're missing is bigint. Which after reading about doesn't seem to important. WDYT?

File: packages/logging/src/internal/pino.ts [hunk 7]
```diff
@@ -71,14 +70,31 @@ const formatError = (input: FormatterInput): string => [
...
+  )
+  const formattedNonPrimitiveExcessArgs = baseExcessArgs
+    .filter(x => !isPrimitiveType(x))
+    .reduce(
+      (total, curr) => ({ ...(total as LogTags), ...(curr as LogTags) }),
+      {}
+    ) as LogTags
```
[the remark this answers is not in the document]
Comment (amitgoldarb) [reply packages/logging/src/internal/pino.ts:92]: refactored and looks better (using Object.assign)

File: packages/logging/src/internal/pino.ts [hunk 8]
```diff
@@ -71,14 +70,31 @@ const formatError = (input: FormatterInput): string => [
...
-  `arg${i}`,
-  isPrimitiveType(value) ? formatLogTagValue(value) : safeStringify(value),
-]
-
-const formatExcessArgs = (excessArgs?: unknown[]): LogTags =>
-  Object.fromEntries((excessArgs || []).map(formatExcessArg))
+const formatPrimitiveExcessArg = (value: unknown, i: number): [string, PrimitiveType] =>
+  (isPrimitiveType(value)
+    ? [`arg${i}`, typeof value === 'string' ? stringifyIfNecessary(value) : value]
+    : ['', '']
```
[the remark this answers is not in the document]
Comment (amitgoldarb) [reply packages/logging/src/internal/pino.ts:76]: refactored and looks better (using Object.assign)

File: packages/logging/src/internal/pino.ts (hunk 1 above)
Comment (amitgoldarb) [reply packages/logging/src/internal/pino.ts:85]: refactored and looks better (using Object.assign). This isn't here anymore

File: packages/logging/test/internal/pino_logger.test.ts (hunk 2 above)
Comment (amitgoldarb) [reply packages/logging/test/internal/pino_logger.test.ts:676]: I'll add some more 👍

File: packages/logging/src/internal/log-tags.ts [hunk 9]
```diff
@@ -31,23 +32,40 @@ const stringifyIfNecessary = (s: string): string => {
...
-  if (typeof value === 'string') return `${stringifyIfNecessary(value)}`
-  return ''
-}
+const encapsulateIfNecessary = (s: string): string => (s.match(/^".*"$/) ? s : `"${s}"`)
 
 export const isPrimitiveType = (value: unknown): value is string | number | boolean =>
   ['string', 'number', 'boolean'].includes(typeof value)
 
+export const formatPrimitiveLogTagValue = (value: unknown): string => {
+  if (typeof value === 'number' || typeof value === 'boolean') return value.toString()
```
[the remark this answers is not in the document]
Comment (amitgoldarb) [reply packages/logging/src/internal/log-tags.ts:41]: Fixed. The compiler didn't understand that value was of a specific type when using switch case. So had to stick with if.

**Review by royra — APPROVED** [review]

**Pull request merged**